src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
I still have real concerns about a pool, even though I appreciate that you've put some effort into addressing my concerns about the lock.

There's still synchronization here.  There's additional allocation: concurrent stack does add additional allocations, and while you've done some profiling and testing, the cost of such small allocations is most impactful on a machine with lots of load and other things causing GCs and the like.  Etc.

There's also still the issue that it seems we have a mismatch between usage and what's being pooled.  The nice thing about StringBuilderCache is that every thread (and only threads) that does these kinds of manipulations gets its own StringBuilder.  As such, we can think of the extra StringBuilder as just being a bit more fixed cost to the already high memory cost of a thread, e.g. instead of ~1MB for the thread, it's ~1MB + a small StringBuilder object.  If you have 1 thread in the application doing this kind of work, you'll never have more than 1 of these cached builders.  If you have 150 threads, you'll have 150 of these, but again it's entirely proportional.  If you had 150 threads and then it went down to 3, you'll end up with only 3 of these.  With the pooling solution, we could both have not enough cached buffers (e.g. 150 threads but we hit the 64 limit in the pool) and too many cached buffers (e.g. we had a brief period of stress that caused all 64 buffers to be created, and then the system quiesced and we're holding on to way more than we actually need).

Longer term, a pooling solution may be reasonable, but I'd rather see us stick with something that's worked well for us in the past, and use a TLS-based solution like we have with StringBuilder.  Essentially then we're just saying that we're switching to some type besides StringBuilder because we don't like its marshaling behavior, and not complicating things with a pool initially. If later we realize that a pool would have other benefits, we could switch to such a thing, but hopefully that wouldn't be until we actually have a built-in pooling solution in the framework that we could utilize here rather than building our own per assembly.

You mentioned in a separate comment that the TLS approach wouldn't work because we may need 2 at a time rather than 1 at a time.  I don't understand why that prevents the TLS approach.  You could have multiple [ThreadLocal] fields.  You could have a single [ThreadLocal] field storing a pair of buffers.  Etc.

Part of my concern may be that it seems we're trying to bite off a ton of changes in a single PR and without actually measuring the impact and benefit of each individual change:
1) We're switching from StringBuilder to another type with different marshaling characteristics.
2) We're using native memory rather than managed memory in that other type, in order to avoid pinning costs as part of marshaling.
3) We're jumping to P/Invokes for the memory allocation rather than .NET's built-in functions for native memory allocation.
4) We're changing the caching mechanism currently being employed.
etc.

I'm suggesting that we stick to just (1) for this change; I think that's reasonable, and it maps to the title of your PR.  I don't believe that implies anything about doing (2), (3), or (4).  We can start with (1), and then in a measured approach can add (2), (3), and/or (4) if they prove to be a real improvement.  Note that (2), (3), and (4) all have cons to them that need to be considered, e.g. using native memory instead of managed means more unsafe code, more SafeHandles, etc., and I've not seen perf measurements that show the pinning to be problematic.